datafusion/core/src/physical_optimizer/pruning.rs

```diff
@@ -495,21 +501,40 @@ fn rewrite_expr_to_prunable(
     }
 
     match column_expr {
+        // col op lit()
+        Expr::Column(_) => Ok((column_expr.clone(), op, scalar_expr.clone())),
         // `cast(col) op lit()`
         Expr::Cast { expr, data_type } => {
-            let (left, op, right) = rewrite_expr_to_prunable(expr, op, scalar_expr)?;
-            Ok((cast(left, data_type.clone()), op, right))
+            let from_type = expr.get_type(&schema)?;
+            if support_type_for_prune(&from_type, data_type) {
+                let (left, op, right) =
+                    rewrite_expr_to_prunable(expr, op, scalar_expr, schema)?;
+                Ok((cast(left, data_type.clone()), op, right))
+            } else {
+                return Err(DataFusionError::Plan(format!(
+                    "Cast with from type {} to type {} is not supported",
+                    from_type, data_type
+                )));
+            }
```

Comment:
We could probably avoid the duplication of the error generation by moving the error return into `support_type_for_prune` like:

```suggestion
            verify_support_type_for_prune(&from_type, data_type)?;
            let (left, op, right) = rewrite_expr_to_prunable(expr, op, scalar_expr)?;
            Ok((cast(left, data_type.clone()), op, right))
```

This is not needed for correctness, just a code structure suggestion

Reply:
good point